File: client/src/main/java/su/plo/voice/client/audio/source/ClientStaticSource.java
package su.plo.voice.client.audio.source;

import org.jetbrains.annotations.NotNull;
import su.plo.voice.api.client.PlasmoVoiceClient;
import su.plo.voice.client.config.ClientConfig;
import su.plo.voice.proto.data.audio.source.StaticSourceInfo;

public final class ClientStaticSource extends BaseClientAudioSource<StaticSourceInfo> {

    public ClientStaticSource(@NotNull PlasmoVoiceClient voiceClient,
                              @NotNull ClientConfig config) {
        super(voiceClient, config);
    }

    @Override
    protected float[] getPosition(float[] position) {
        position[0] = (float) sourceInfo.getPosition().getX();
        position[1] = (float) sourceInfo.getPosition().getY();
        position[2] = (float) sourceInfo.getPosition().getZ();

        return position;
    }

    @Override
    protected float[] getLookAngle(float[] lookAngle) {
        lookAngle[0] = (float) sourceInfo.getLookAngle().getX();
        lookAngle[1] = (float) sourceInfo.getLookAngle().getY();
        lookAngle[2] = (float) sourceInfo.getLookAngle().getZ();

        return lookAngle;
    }
}

File: client/src/main/kotlin/su/plo/voice/client/audio/source/ClientStaticSource.kt
package su.plo.voice.client.audio.source

import net.minecraft.world.phys.Vec3
import su.plo.slib.api.position.Pos3d
import su.plo.voice.client.BaseVoiceClient
import su.plo.voice.client.config.VoiceClientConfig
import su.plo.voice.client.extension.toVec3
import su.plo.voice.proto.data.audio.source.StaticSourceInfo

class ClientStaticSource(
    voiceClient: BaseVoiceClient,
    config: VoiceClientConfig,
    sourceInfo: StaticSourceInfo
) : BaseClientAudioSource<StaticSourceInfo>(voiceClient, config, sourceInfo) {

    val lastRenderPosition: Pos3d = sourceInfo.position.copy()

    override fun getPosition(): Vec3 =
        sourceInfo.position.toVec3()

    override fun getLookAngle(): Vec3 =
        sourceInfo.lookAngle.toVec3()

    override fun shouldCalculateDirectionalGain() =
        sourceInfo.angle > 0 && super.shouldCalculateDirectionalGain()
}
